### PR TITLE
[SPARK-44640][PYTHON] Improve error messages for Python UDTF returning non Iterable

### PR DESCRIPTION
**Pull request by allisonwang-db**
### What changes were proposed in this pull request?

This PR improves error messages when the result of a Python UDTF is not an Iterable. It also improves the error messages when a UDTF encounters an exception when executing `eval`.

### Why are the changes needed?

To make Python UDTFs more user-friendly.

### Does this PR introduce _any_ user-facing change?

Yes. For example this UDTF:
```
@udtf(returnType="x: int")
class TestUDTF:
    def eval(self, a):
        return a 
```
Before this PR, it fails with this error for regular UDTFs:
```
    return tuple(map(verify_and_convert_result, res))
TypeError: 'int' object is not iterable
```
And this error for arrow-optimized UDTFs:
```
    raise ValueError("DataFrame constructor not properly called!")
ValueError: DataFrame constructor not properly called!
```

After this PR, the error message will be:
`pyspark.errors.exceptions.base.PySparkRuntimeError: [UDTF_RETURN_NOT_ITERABLE] The return value of the UDTF is invalid. It should be an iterable (e.g., generator or list), but got 'int'. Please make sure that the UDTF returns one of these types.`

### How was this patch tested?

New UTs.

### CODE REVIEW

**Comment by allisonwang-db** [comment]
cc @ueshin @HyukjinKwon

File: python/pyspark/worker.py [hunk 1]
```diff
@@ -656,32 +651,46 @@ def mapper(_, it):
         def wrap_udtf(f, return_type):
             assert return_type.needConversion()
             toInternal = return_type.toInternal
+            return_type_size = len(return_type)
 
             def verify_and_convert_result(result):
                 # TODO(SPARK-44005): support returning non-tuple values
                 if result is not None and hasattr(result, "__len__"):
-                    if len(result) != len(return_type):
+                    if len(result) != return_type_size:
```
Comment (allisonwang-db) [inline python/pyspark/worker.py:659]: This is to address this comment: https://github.com/apache/spark/pull/42157/files#r1276758240

**Review by ueshin — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.5.

**Comment by HyukjinKwon** [comment]
It has a conflict with branch-3.5. Should we create a PR for it?

**Comment by allisonwang-db** [comment]
Yup we need this in branch-3.5. Created https://github.com/apache/spark/pull/42337

**Pull request closed without merging**